File: services/llm-gateway/src/gateway.py
from __future__ import annotations

import os
from pathlib import Path
from typing import Any, Callable

from llm_logger import log_llm_call
from model_router import route_model
from prompt_guard import validate_llm_request
from provider_deepseek import Transport, chat_completion
from schemas import EvidenceChunk, LLMRequest


def llm_enabled() -> bool:
    return os.environ.get("DEEPSEEK_ENABLE_LLM") == "1" and bool(os.environ.get("DEEPSEEK_API_KEY"))


def generate_admissions_answer(
    *,
    project_root: Path,
    request: LLMRequest,
    transport: Transport | None = None,
) -> str | None:
    if not llm_enabled() and transport is None:
        return None

    policy_request = request.to_policy_dict()
    ok, violations = validate_llm_request(policy_request)
    if not ok:
        log_llm_call(
            project_root,
            {
                "status": "blocked",
                "blocked_by": "prompt_guard",
                "violations": violations,
                "request": policy_request,
            },
        )
        return None

    route = route_model("admissions_answer", {"role": request.user_role})
    messages = _build_messages(policy_request)
    try:
        answer = chat_completion(
            messages,
            model=route["model"],
            transport=transport,
        )
    except Exception as exc:
        log_llm_call(
            project_root,
            {
                "status": "error",
                "route": route,
                "error": str(exc),
                "request": policy_request,
            },
        )
        return None

    log_llm_call(
        project_root,
        {
            "status": "ok",
            "route": route,
            "request": policy_request,
            "answer": answer,
        },
    )
    return answer


def _build_messages(request: dict[str, Any]) -> list[dict[str, str]]:
    scope = request["scope"]
    system = (
        "你是复读学校招生问答 Agent。只能使用提供的 allowed evidence 回答。"
        "不要编造政策、价格、名额或录取结果。不得承诺固定提分或保证录取。"
        "如果角色是 visitor/student/parent，不得输出 internal/L3/L4 内容。"
        "如果角色是 sales 且证据含 internal 内容，必须标注“内部参考”，并提醒不要原样对外发送。"
        "回答末尾用“来源：”列出证据标题。"
    )
    user = (
        f"用户角色：{scope['role']}\n"
        f"校区：{scope.get('campus')}\n"
        f"意图：{request['intent']}\n"
        f"用户问题：{request['message']}\n"
        f"allowed evidence：{request['evidence']}\n"
    )
    return [
        {"role": "system", "content": system},
        {"role": "user", "content": user},
    ]


def _clean_chunk(content: str) -> str:
    lines = []
    for line in content.splitlines():
        stripped = line.strip()
        if stripped.startswith("#"):
            continue
        if stripped:
            lines.append(stripped)
    return " ".join(lines)

File: services/llm-gateway/ts-src/gateway.ts
import { LLMRequestSchema, llmRequestToPolicyDict } from "./schemas.js";
import { validateLlmRequest } from "./promptGuard.js";
import { routeModel } from "./modelRouter.js";
import { chatCompletion } from "./providerDeepseek.js";
import { logLlmCall } from "./llmLogger.js";
import { llmEnabled } from "./config.js";
import type { Transport } from "./providerDeepseek.js";

function pyRepr(val: unknown): string {
  if (val === null) return "None";
  if (typeof val === "string") return `'${val}'`;
  if (typeof val === "number" || typeof val === "boolean") return String(val);
  if (Array.isArray(val)) {
    const items = val.map(v => pyRepr(v));
    return `[${items.join(", ")}]`;
  }
  if (typeof val === "object") {
    const obj = val as Record<string, unknown>;
    const entries = Object.entries(obj);
    const inner = entries.map(([k, v]) => `'${k}': ${pyRepr(v)}`);
    return `{${inner.join(", ")}}`;
  }
  return String(val);
}

function buildMessages(request: Record<string, unknown>): Array<{ role: string; content: string }> {
  const scope = (request["scope"] ?? {}) as Record<string, unknown>;
  const system = "你是复读学校招生问答 Agent。只能使用提供的 allowed evidence 回答。不要编造政策、价格、名额或录取结果。不得承诺固定提分或保证录取。如果角色是 visitor/student/parent，不得输出 internal/L3/L4 内容。如果角色是 sales 且证据含 internal 内容，必须标注\u201c内部参考\u201d，并提醒不要原样对外发送。回答末尾用\u201c来源：\u201d列出证据标题。";
  const role = "role" in scope ? String(scope["role"] ?? "None") : "None";
  const campus = "campus" in scope ? String(scope["campus"] ?? "None") : "None";
  const intent = "intent" in request ? String(request["intent"] ?? "None") : "None";
  const message = "message" in request ? String(request["message"] ?? "None") : "None";
  const evidence = "evidence" in request ? pyRepr(request["evidence"]) : "None";
  const user = `用户角色：${role}\n校区：${campus}\n意图：${intent}\n用户问题：${message}\nallowed evidence：${evidence}\n`;
  return [
    { role: "system", content: system },
    { role: "user", content: user },
  ];
}

export async function generateAdmissionsAnswer(
  projectRoot: string,
  request: Record<string, unknown>,
  transport?: Transport,
): Promise<string | null> {
  if (!llmEnabled() && transport === undefined) {
    return null;
  }

  const parsed = LLMRequestSchema.parse(request);
  const policyRequest = llmRequestToPolicyDict(parsed);
  const { valid, violations } = validateLlmRequest(policyRequest);
  if (!valid) {
    logLlmCall(projectRoot, {
      status: "blocked",
      blocked_by: "prompt_guard",
      violations,
      request: policyRequest,
    });
    return null;
  }

  const route = routeModel("admissions_answer", { role: parsed.user_role });
  const messages = buildMessages(policyRequest);

  let answer: string;
  try {
    answer = await chatCompletion(messages, {
      model: route["model"]!,
      transport,
    });
  } catch (exc: unknown) {
    const errMsg = exc instanceof Error ? exc.message : String(exc);
    logLlmCall(projectRoot, {
      status: "error",
      route,
      error: errMsg,
      request: policyRequest,
    });
    return null;
  }

  logLlmCall(projectRoot, {
    status: "ok",
    route,
    request: policyRequest,
    answer,
  });
  return answer;
}
